lsp-mode.el

```diff
@@ -4659,14 +4659,68 @@ Stolen from `org-copy-visible'."
 
     (lsp--setup-markdown major-mode)))
 
+(defvar lsp--display-image-alist
+  '((lsp--render-markdown
+     (:regexp
+      "!\\[.*?\\](data:image/png;base64,\\([A-Za-z0-9+/\n]+?=*?\\)\\(|[^)]+\\)?)"
+      :sexp
+      (create-image
+       (base64-decode-string
+        (buffer-substring-no-properties (match-beginning 1) (match-end 1)))
+       nil t))))
+  "Replaced string regexp and function returning image.
+Each element should be:
+(MODE . (PROPERTY-LIST...))
+MODE (car) is function which is defined in `lsp-language-id-configuration'.
+Cdr should be list of PROPERTY-LIST.
+
+Each PROPERTY-LIST should have properties:
+:regexp  Regexp which determines what string is relpaced to image.
+         You should also get infomation of image, by parenthesis constructs.
+         By deafult, all matched string is replaced to image, but you can change
+         index of replaced string by keyword :replaced-index.
+
+:sexp    Return image when evaluated. You can use infomation of regexp
+         by using (match-beggining N), (match-end N) or (match-substring N).
+
+Inaddition, Each can have property:
+:replaced-index  Determine index which is used to replace regexp to image.
+                 The value means first argument of `match-beggining' and `match-end'.
+                 If omitted, interpreted as index 0.
+")
+
+(defun lsp--display-image (mode)
+  "Add image property if available."
+  (let ((plist-list (cdr (assq mode lsp--display-image-alist))))
+    (when (display-images-p)
+      (cl-loop
+       for plist in plist-list
+       with regexp with replaced-index
+       do
+       (setq regexp (plist-get plist :regexp))
+       (setq replaced-index (or (plist-get plist :replaced-index) 0))
+
+       (font-lock-remove-keywords nil (list regexp replaced-index))
+       (let ((inhibit-read-only t))
+         (save-excursion
+           (goto-char (point-min))
+           (while (re-search-forward regexp nil t)
+             (set-text-properties
+              (match-beginning replaced-index) (match-end replaced-index)
+              nil)
+             (add-text-properties
+              (match-beginning replaced-index) (match-end replaced-index)
+              `(display ,(eval (plist-get plist :sexp)))))))))))
+
 (defun lsp--fontlock-with-mode (str mode)
   "Fontlock STR with MODE."
   (condition-case nil
       (with-temp-buffer
         (insert str)
         (delay-mode-hooks (funcall mode))
         (cl-flet ((window-body-width () lsp-window-body-width))
-          (font-lock-ensure))
+          (font-lock-ensure)
+          (lsp--display-image mode))
```

Comment:
Seems like this should go in lsp--render-markdown ?

Reply:
@yyoncho 
As far as I have experimented, it might be impossible, because `font-lock-ensure` rewrites or overwrites the text property as an link. So there are no choice except for putting displaying function after `font-lock-ensure`,  I think.

In addition, putting it to `lsp--render-markdown` seems to make showing image difficult to extend to other modes, or other image types.

Reply:
I can make `markdown-toggle-inline-images`, but anyway, `lsp--display-image` must identify mode as `lsp--render-markdown` and be called after `font-lock`. There are almost no differrence whether markdown-specific or not. 
Toggling function or variable should exist, I think, too. 

And, `lsp-display-image` cannot deal with images except base64 in markdown yet, so it should be extendable, I think.